File: src/main/java/net/ndrei/teslacorelib/items/RegisteredItem.java
package net.ndrei.teslacorelib.items;

import com.google.common.collect.Lists;
import net.minecraft.client.renderer.block.model.ModelResourceLocation;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.Item;
import net.minecraft.item.crafting.CraftingManager;
import net.minecraft.item.crafting.IRecipe;
import net.minecraftforge.client.model.ModelLoader;
import net.minecraftforge.fml.common.registry.GameRegistry;

import java.util.List;

/**
 * Created by CF on 2016-12-13.
 */
public class RegisteredItem extends Item {
    public RegisteredItem(String modId, CreativeTabs tab, String registryName) {
        super();

        this.setRegistryName(modId, registryName);
        this.setUnlocalizedName(modId + "_" + registryName);
        if (tab != null) {
            this.setCreativeTab(tab);
        }
    }

    public void register() {
        GameRegistry.register(this);

        List<IRecipe> recipes = this.getRecipes();
        if (recipes != null) {
            recipes.forEach(recipe -> {
                if (recipe != null) {
                    CraftingManager.getInstance().addRecipe(recipe);
                }
            });
        }
    }

    protected IRecipe getRecipe() {
        return null;
    }

    protected List<IRecipe> getRecipes() {
        List<IRecipe> recipes = Lists.newArrayList();

        IRecipe recipe = this.getRecipe();
        if (recipe != null)
            recipes.add(recipe);

        return recipes;
    }

    public void registerRenderer() {
        ModelLoader.setCustomModelResourceLocation(
                this,
                0,
                new ModelResourceLocation(this.getRegistryName(), "inventory")
        );
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/items/RegisteredItem.kt
package net.ndrei.teslacorelib.items

import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.Item
import net.minecraft.item.crafting.IRecipe
import net.minecraft.util.ResourceLocation
import net.minecraftforge.client.model.ModelLoader
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import net.minecraftforge.registries.IForgeRegistry

/**
 * Created by CF on 2017-06-22.
 */
abstract class RegisteredItem(modId: String, tab: CreativeTabs?, registryName: String) : Item() {
    init {
        this.setRegistryName(modId, registryName)
        this.unlocalizedName = modId + "." + registryName
        if (tab != null) {
            this.creativeTab = tab
        }
    }

    open fun register(registry: IForgeRegistry<Item>) {
        registry.register(this)
    }

    fun registerRecipe(registry: (recipe: IRecipe) -> ResourceLocation)
            = this.recipes.forEach { registry(it) }

    protected open val recipe: IRecipe?
        get() = null

    protected open val recipes: List<IRecipe>
        get() {
            val recipe = this.recipe
            return if (recipe != null) listOf(recipe) else listOf()
        }

    @SideOnly(Side.CLIENT)
    open fun registerRenderer() = ModelLoader.setCustomModelResourceLocation(this, 0, ModelResourceLocation(this.registryName!!, "inventory"))
}
